File: tools/plot/matlab/read_ev_file.m
function [ nuc_name, xmf, time, temperature, density, timestep, edot] = read_ev_file ( ev_filename )
%--------------------------------------------------------------------------
%[ nuc_name, xmf, time, temperature, density, timestep, edot] = read_ev_file ( ev_filename ) 
% Reads XNet ASCII output file.
% Inputs>  ev_filename: name of ASCII file 
% Outputs< nuc_name: symbol for each isotope in the file.
%          xmf: time series of mass fractions 
%          time: discrete temporal evolution
%          temperature:  time series of temperature
%          density: time series of density
%          timestep:  time series of discrete timesteps 
%          edot: time series of energy generation 
%--------------------------------------------------------------------------

% Set number of species
  ny=14

% Report the absence of flux data
  nflx=0

% Read Header
  label=textread(ev_filename,'%s',ny+7);

% Read data from file as block
  raw_data=textread(ev_filename,'%f','headerlines', 1);

% Reshape data into columns
  data_size=size(raw_data);
  read_width=ny+8;
  read_length=data_size(1)/read_width;
  read_data=reshape(raw_data,read_width,read_length);

% Extract variables from columns
  k=read_data(1,:);
  time=read_data(2,:);
  temperature=read_data(3,:);
  density=read_data(4,:);
  edot=read_data(5,:);
  timestep=read_data(6,:);
  xmf=read_data(7:6+ny,:);

% Reformat isotope symbols
  zz=regexprep(label(7:6+ny),'\d+','');
  aa=regexprep(label(7:6+ny),'\D+','');
  zz=regexprep(zz,'(^.)','${upper($1)}');
  nuc_name=strcat('^{',aa,'}',zz);

% Calculate # of timesteps
nstep = size(time,2)

end

File: tools/python/read_ev_file.py
"""
    Module to read ev files
    """
def read_ev_file(datafile):
    """
        Inputs > datafile, name of input file (ev file)
        Outputs< time, elapsed time as a function of time
        temperature, temperature (in GK) as a function of time
        density, mass density (in g/cm3) as a function of time
        timestep, timestep as a function of time
        edot, energy generation rate as a function of time
        species, a list of each species
        data, a shortcut to np.genfromtxt that reads the given datafile
        datafile, the same ev file from input
        """
    import numpy as np
    
    #Read the data file
    col_headers = np.genfromtxt(datafile, max_rows = 1, dtype = str)
    
    if len(col_headers) != 22:
        col_headers[20] = 'I'
        col_headers = np.append(col_headers, 't')
        col_headers_list = col_headers.tolist()
        data = np.genfromtxt(datafile, skip_header = 1, dtype = float, names = col_headers_list)
    
    else:
        data = np.genfromtxt(datafile, names=True)
    
    #Sort columns of data by name. Store in a list.
    columns = []
    for name in data.dtype.names:
        columns.append(name)
    
    #Save columns as variables.
    time = data['Time']
    density = data['Density']
    temperature = data['TGK']
    edot = data['dEdt']
    timestep = data['Timestep']

    #Make a list of species mass fraction data (xmf).
    species = []
    alphabet = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'v', 'w', 'x', 'y', 'z']
    
    for item in columns:
        if item != 'Time' and item != 'TGK' and item != 'Density' and item != 'dEdt' and item != 'Timestep':
            species.append(item)

    for item in species:
        if item in alphabet:
            species.remove(item)
    
    return time, density, temperature, edot, timestep, species, data, datafile

